assembly/bindgen.ts

```diff
@@ -233,6 +272,20 @@ function decode<T, V = Uint8Array>(buf: V, name: string = ""): T {
     //@ts-ignore
     return value.decode<JSON.Obj>(<JSON.Obj>val);
   }
+  //@ts-ignore
+  if (value instanceof Map) {
+    assert(val instanceof JSON.Obj, "Value with Key: " + name + " of type map expected a JSON.Obj, but recevied " + JSONTypeToString(val));
+    //@ts-ignore
+    assert(nameof<indexof<T>>() == "String", "Value with Key: " + name + " cannot decode a map which has an index type "+ nameof<indexof<T>>() + ", it must be a string");
+    //@ts-ignore
+    return <T>decodeMap<valueof<T>>(<JSON.Obj>val, name);
+  }
+  //@ts-ignore
+  if (value instanceof Set) {
+    assert(val instanceof JSON.Arr, "Value with Key: " + name + " of type map expected a JSON.Obj, but recevied " + JSONTypeToString(val));
+    //@ts-ignore
```

Comment:
Why are all the `@ts-ignore` comments needed? They're so frequent in this PR that it makes it seem like we have a dependency that is no longer serving us

Reply:
lol it's to appease the Typescript compiler.  For example, in the line below we know that casting the return value with `<T>` is correct, but typescript doesn't because it doesn't have any info on the type, e.g. at the top `<T extends { toString(): string }>`.

This will get better once AS has it's own linter/IDE support.